File: python-design-patterns-2017/examples/example9/Strategy.java
class Customer {

    private List<Double> drinks;
    private BillingStrategy strategy;

    public Customer(final BillingStrategy strategy) {
        this.drinks = new ArrayList<Double>();
        this.strategy = strategy;
    }

    public void add(final double price, final int quantity) {
        drinks.add(strategy.getActPrice(price*quantity));
    }

    // Payment of bill
    public void printBill() {
        double sum = 0;
        for (Double i : drinks) {
            sum += i;
        }
        System.out.println("Total due: " + sum);
        drinks.clear();
    }

    // Set Strategy
    public void setStrategy(final BillingStrategy strategy) {
        this.strategy = strategy;
    }

}

interface BillingStrategy {
    double getActPrice(final double rawPrice);
}

// Normal billing strategy (unchanged price)
class NormalStrategy implements BillingStrategy {

    @Override
    public double getActPrice(final double rawPrice) {
        return rawPrice;
    }

}

// Strategy for Happy hour (50% discount)
class HappyHourStrategy implements BillingStrategy {

    @Override
    public double getActPrice(final double rawPrice) {
        return rawPrice*0.5;
    }

}

// Source: https://en.wikipedia.org/wiki/Strategy_pattern

File: python-design-patterns-2017/examples/example9/strategy.py
class Customer:
    def __init__(self, strategy):
        self.drinks = []
        self.strategy = strategy
    
    def add(self, price, quantity):
        drinks.append(strategy(price * quantity))

    def print_bill(self):
        ...

# strategy is just a function that returns a price

c = Customer(lambda x: x)
c.add(5.50, 2)
c.add(4.30, 1)
# Happy hour
c.strategy = lambda x: x * 0.5
c.add(4.30, 6)
